Accept UnaryMinusExpression as class parameter type

joshcooper opened the pull request:
Previously, class parameters of the form `Integer[-1] $param` would fail
compilation, because the value `-1` was lexed as a UnaryMinusExpression
containing a LiteralInteger. And since the LiteralEvaluator didn't implement the
`literal_UnaryMinusExpression` method, the visitor called `literal_XXX` for each
ancestor class, until reaching `literal_Object`, which always raises.

This adds the `literal_UnaryMinusExpression` method and returns -1 times
the expression it wraps. This is similar to how the TypeParser interprets
UnaryMinusExpressions[1].

If strict is off, the issue will be ignored. If strict is warning, then a
warning will be reported, but compilation will continue:

    Warning: The parameter '$i' must be a literal type, not a Puppet::Pops::Model::AccessExpression

If strict is error, then compilation will fail.

Fixes #9268

[1] https://github.com/puppetlabs/puppet/blob/8.5.0/lib/puppet/pops/types/type_parser.rb#L161

joshcooper commented on spec/unit/pops/evaluator/literal_evaluator_spec.rb:
```diff
@@ -37,7 +38,7 @@
   end
 
-  ['1+1', '[1,2, 1+2]', '{a=>1+1}', '"x$y"', '"x${y}z"', 'Integer[1-3]', 'Optional[[String]]'].each do |source|
+  ['1+1', '[1,2, 1+2]', '{a=>1+1}', '"x$y"', '"x${y}z"', 'Integer[1-3]', 'Intger[-1-3]', 'Optional[[String]]'].each do |source|
```
Integer typo

donoghuc commented on lib/puppet/pops/validation/validator_factory_4_0.rb:
```diff
@@ -36,6 +36,8 @@ def severity_producer
     p[Issues::EMPTY_RESOURCE_SPECIALIZATION]  = :ignore
     p[Issues::CLASS_NOT_VIRTUALIZABLE]        = :error
+
+    p[Issues::ILLEGAL_NONLITERAL_PARAMETER_TYPE] = Puppet[:strict] == :off ? :ignore : Puppet[:strict]
```
Seems like the least surprising thing would be to let it default to `warning`. In the case there is just a warning would the type just not be enforced? That doesn't seem terribly dangerous?

joshcooper replied:
It's a bit subtle. In the case of `Integer[1-3]`, the type is actually valid, because it evaluates to `Integer[-2]`, which in the context of a class parameter `Integer[-2] $i` means it will accept integers >= -2. However, the manifest author intended to specify the range `Integer[1, 3]`. So if you pass a value bigger than 3, then it passes validation (surprise)!

donoghuc reviewed (approved):
Seems good. Double checked ranges work as expected locally 

```diff
diff --git a/spec/unit/pops/evaluator/literal_evaluator_spec.rb b/spec/unit/pops/evaluator/literal_evaluator_spec.rb
index 63ee699a78..90a0aeb201 100644
--- a/spec/unit/pops/evaluator/literal_evaluator_spec.rb
+++ b/spec/unit/pops/evaluator/literal_evaluator_spec.rb
@@ -18,6 +18,8 @@ describe "Puppet::Pops::Evaluator::LiteralEvaluator" do
     'a::b'    => 'a::b',
     'Integer[1]' => [1],
     'Integer[-1]' => [-1],
+    'Integer[-5, -1]' => [-5, -1],
+    'Integer[-5, 5]' => [-5, 5],
     'File' => "file",
```

I like that it will follow `strict` settings. So effectively it will default to `warning` in 7 and `error` in 8. 👍

evgeni commented on spec/unit/pops/evaluator/literal_evaluator_spec.rb:
```diff
@@ -11,12 +11,32 @@
     "3.14"    => 3.14,
     "true"    => true,
     "false"   => false,
+    "undef"   => nil,
```
this duplicates the later `it "evaluates undef to nil" do` test

rembun commented:
I am getting this error on fresh install foreman running on alma 8  foreman-installer  [root@localhost ~]# vi  /usr/share/foreman-installer/modules/redis/manifests/init.pp

/usr/share/gems/gems/kafo-7.4.0/lib/kafo/puppet_module.rb:70:in `parse': No Puppet module parser is installed and no cache of the file /usr/share/foreman-installer/modules/foreman/manifests/cli.pp is available. Please check debug logs and install optional dependencies for the parser. (Kafo::ParserError)
        from /usr/share/gems/gems/kafo-7.4.0/lib/kafo/configuration.rb:133:in `block in modules'
        from /usr/share/gems/gems/kafo-7.4.0/lib/kafo/configuration.rb:133:in `map'
        from /usr/share/gems/gems/kafo-7.4.0/lib/kafo/configuration.rb:133:in `modules'
        from /usr/share/gems/gems/kafo-7.4.0/lib/kafo/kafo_configure.rb:257:in `modules'
        from /usr/share/gems/gems/kafo-7.4.0/lib/kafo/kafo_configure.rb:261:in `module'
        from /usr/share/gems/gems/kafo-7.4.0/lib/kafo/hook_context.rb:113:in `module_present?'
        from /usr/share/foreman-installer/hooks/boot/20-certs_update.rb:2:in `block (4 levels) in load'
        from /usr/share/gems/gems/kafo-7.4.0/lib/kafo/hooking.rb:36:in `instance_eval'
        from /usr/share/gems/gems/kafo-7.4.0/lib/kafo/hooking.rb:36:in `block (4 levels) in load'
        from /usr/share/gems/gems/kafo-7.4.0/lib/kafo/hook_context.rb:19:in `instance_eval'
        from /usr/share/gems/gems/kafo-7.4.0/lib/kafo/hook_context.rb:19:in `execute'
        from /usr/share/gems/gems/kafo-7.4.0/lib/kafo/hooking.rb:67:in `block in execute'
        from /usr/share/gems/gems/kafo-7.4.0/lib/kafo/hooking.rb:65:in `each'
        from /usr/share/gems/gems/kafo-7.4.0/lib/kafo/hooking.rb:65:in `execute'
        from /usr/share/gems/gems/kafo-7.4.0/lib/kafo/kafo_configure.rb:144:in `initialize'
        from /usr/share/gems/gems/clamp-1.3.2/lib/clamp/command.rb:140:in `new'
        from /usr/share/gems/gems/clamp-1.3.2/lib/clamp/command.rb:140:in `run'
        from /usr/share/gems/gems/kafo-7.4.0/lib/kafo/kafo_configure.rb:54:in `run'
        from /usr/sbin/foreman-installer:8:in `<main>'

rembun commented:
Has the changed already pushed to the repo as it seems that we still pulling wrong code

mhashizume commented:
@rembun We released Puppet 7.29.1 and 8.5.1 today with this fix included. Please let us know if those versions do not resolve this issue for you.

kjetilho commented:
This fix is not sufficient (I tested using 7.29.1. gem), it breaks arithmetic in general.  Examples from my code:

```puppet
  Integer[0, 3*3600] $random_delay = 1800,
  Integer[0, 24*60] $downtime_duration = 10,
```

ekohl commented:
@donoghuc isn't that an unusual thing to do in a minor release?

joshcooper commented:
@kjetilho please file a new GH issue instead of reusing this one. In the new issue, can you explain what you mean by "it breaks arithmetic in general"? How exactly is it broken? What behavior do you see?

kjetilho commented:
@joshcooper - I included both my code and the error message.  What more do you need?

@donoghuc - thanks, I see it was intended to break my code.  In which case I must agree with @ekohl, that seems like an unusual thing to do in a minor release, and at least warrants an entry in the release notes.

In any case, the goal of "provid[ing] a descriptive warning if parameter(s) are not literal" was not met IMHO.  It is not descriptive for users (no users have any idea what an `Puppet::Pops::Model::AccessExpression` is), and it is not a warning :smile:

joshcooper commented:
@kjetilho you said, "This fix is not sufficient (I tested using 7.29.1. gem), it breaks arithmetic in general" And you gave the resulting message, "The parameter '$random_delay' must be a literal type, not a Puppet::Pops::Model::AccessExpression" However, isn't it true that in 7.29.1 the message is printed as a warning in the puppetserver logs? If so, that's very different than "it breaks".

kjetilho commented:
Ah, I am sorry - my pipeline failed on the first step, which was `rake syntax` ('puppet-syntax/tasks/puppet-syntax').  That's where it was a fatal error.  You are correct, it is just a warning with default settings for `puppet apply`.

joshcooper commented:
@kjetilho ah makes sense, thank you! When `rake syntax` fails does that cause your CI pipelines to fail too?

I would expect puppet to be able to add warnings to the language runtime and evaluation without breaking all module CI, as described in https://github.com/puppetlabs/puppet/blob/7.29.1/lib/puppet/defaults.rb#L192-L204.

It seems puppet-syntax doesn't distinguish between warnings and errors in manifests https://github.com/voxpupuli/puppet-syntax/blob/8114392547895ef352cfa66c3becdc3cf33859fb/lib/puppet-syntax/manifests.rb#L24-L25

but it does for templates https://github.com/voxpupuli/puppet-syntax/blob/8114392547895ef352cfa66c3becdc3cf33859fb/lib/puppet-syntax/templates.rb#L38-L57

Maybe it should for manifests so it's clear the "must be a literal type" message is a warning and not an error? Thoughts @bastelfreak

kjetilho commented:
Yes, the syntax check is strict, but I think it is good to fix even deprecations early.  The puppet-lint in my CI pipeline, is strict, too.  It even regards trailing whitespace a fatal error :)

To summarise my position: I'm fine with the change in behaviour - but I would love to see the warning message improved to something more easily human understandable if possible.

> The parameter '$random_delay' must be a literal type, not a Puppet::Pops::Model::AccessExpression

This indicates to me that the problem is with my parameter value, not the parameter declaration.  How about:

> The type for parameter '$random_delay' must be a literal type, not an expression

Of course my intuition is not everyone else's intution, so that's just my 2 cents.

joshcooper commented:
> The parameter '$random_delay' must be a literal type, not a Puppet::Pops::Model::AccessExpression

This is a bug, we should be using the `LabelProvider` to format the model. How about?

```
The parameter '$i' must be a literal type, not a '*' expression (line: 1, column: 18)
```

> I think this is a breaking change and should have been landed in Puppet 9 and not in a minor release.

In the past, we have introduced new language validations in minor or patch releases in one of two ways:

First, the issue is introduced as a deprecation. It is not under control of `strict`, so `strict=off` does nothing. All deprecations can be disabled via `disable_warnings=deprecations`. Examples of this include:

* `ILLEGAL_DEFINITION_LOCATION` First introduced in 5.5.6 [under `strict` control](https://github.com/puppetlabs/puppet/commit/6e7835b0ebead45f1ba83f4d8ff27d43224ba77f#diff-d72cbd18b52f7f0be850951a13ec04f02349aa35625dcfe973bfd0b26bd9c1d8) but [converted to a `deprecation` before 5.5.6 was released](https://github.com/puppetlabs/puppet/commit/44c4ea06f5c5fd172dcc50b189548209b8ab890f#diff-d72cbd18b52f7f0be850951a13ec04f02349aa35625dcfe973bfd0b26bd9c1d8). In 6.0, converted to an [`error`](https://github.com/puppetlabs/puppet/commit/81e61de735d9fab90553cd0c4ce457f364e6cb96#diff-d72cbd18b52f7f0be850951a13ec04f02349aa35625dcfe973bfd0b26bd9c1d8)
* `ILLEGAL_TOP_CONSTRUCT_LOCATION` First introduced in 5.5.7 as a [`deprecation`](https://github.com/puppetlabs/puppet/commit/56615099b0d85ed5c93cef3d896b8b90c1dd34e9#diff-d72cbd18b52f7f0be850951a13ec04f02349aa35625dcfe973bfd0b26bd9c1d8). In 6.0, converted to an [`error`](https://github.com/puppetlabs/puppet/commit/81e61de735d9fab90553cd0c4ce457f364e6cb96)

Second, the issue is introduced under control of `strict`. By default, that is a `warning` in 7.x and `error` in 8.x, but can be turned off using `strict=off`. Examples of this include:

* `DUPLICATE_KEY` First introduced in 4.5.0 as a [`warning`](https://github.com/puppetlabs/puppet/commit/7300d47425a7ab87c71fe18bce514ae70daa435b#diff-d72cbd18b52f7f0be850951a13ec04f02349aa35625dcfe973bfd0b26bd9c1d8), but converted to be [under the control of `strict` before 4.5.0 was released](https://github.com/puppetlabs/puppet/commit/135b81cb28b70647d11b79858e9db802621c4afb#diff-d72cbd18b52f7f0be850951a13ec04f02349aa35625dcfe973bfd0b26bd9c1d8)
* `DUPLICATE_DEFAULT` First introduced in 4.9.0 under the [control of `strict`](https://github.com/puppetlabs/puppet/commit/95b3be4c15fecb3eafecc512104265a8b1d4d392#diff-d72cbd18b52f7f0be850951a13ec04f02349aa35625dcfe973bfd0b26bd9c1d8). In 4.10.5, converted to an [error](https://github.com/puppetlabs/puppet/commit/c2d14d8359ac51d65e881f2e6b33b67a0ed4ccd4)
* `CLASS_NOT_VIRTUALIZEABLE` First introduced in 4.10.2 under the [control of `strict`](https://github.com/puppetlabs/puppet/commit/f0f32874805ed584101483771ffc21bb24d166aa#diff-d72cbd18b52f7f0be850951a13ec04f02349aa35625dcfe973bfd0b26bd9c1d8). In 7.0, converted to an [error](https://github.com/puppetlabs/puppet/commit/0befcf3270d738aeb1a480590ee824c82ca312c8#diff-d72cbd18b52f7f0be850951a13ec04f02349aa35625dcfe973bfd0b26bd9c1d8)

The reason we went with the "under control of strict" route is because class parameters were always assumed to be literal and puppet's [`environment_classes` REST API](https://www.puppet.com/docs/puppet/6/server/puppet-api/v3/environment_classes#uses-typed-values) failed when they were non-literal, for example `Integer[0, 2*2]`. So there is some urgency around fixing puppet code, as opposed to an arbitrary decision on our part.

But given the precedent with how we handled validations in 5.x and feedback on this thread, I agree we should switch the `ILLEGAL_NONLITERAL_PARAMETER_TYPE` validation to a `deprecation`. It also means the next releases of 7.x and 8.x will behave consistently. I'll file a new issue for this including the label provider.

bastelfreak commented:
> The parameter '$i' must be a literal type, not a '*' expression (line: 1, column: 18)

sounds good to me!
